swupd/delta.go

```diff
@@ -0,0 +1,160 @@
+package swupd
+
+import (
+	"fmt"
+	"os"
+	"sync"
+
+	"github.com/clearlinux/mixer-tools/helpers"
+	"github.com/pkg/errors"
+)
+
+// CreateDeltas creates all delta files between the previous and current
+// version of the supplied manifest, returning a list of files that
+// delta creation failed for.
+func CreateDeltas(manifest *Manifest, statedir string) ([]*File, error) {
+	var c config
+	c, err := getConfig(statedir)
+	if err != nil {
+		return nil, err
+	}
+
+	var oldManifest *Manifest
+	var newManifest *Manifest
+
+	if oldManifest, err = ParseManifestFile(fmt.Sprintf("%s/%d/%s", c.outputDir, manifest.Header.Previous, manifest.Name)); err != nil {
+		return nil, err
+	}
+	if newManifest, err = ParseManifestFile(fmt.Sprintf("%s/%d/%s", c.outputDir, manifest.Header.Version, manifest.Name)); err != nil {
+		return nil, err
+	}
+
+	_ = newManifest.linkPeersAndChange(oldManifest)
+
+	deltas, err := consolidateDeltaFiles(newManifest, c)
+	if err != nil {
+		return nil, errors.Wrapf(err, "Failed to create deltas list %s", manifest.Name)
+	}
+
+	var failed []*File
+	var errchan = make(chan *File)
+	var wg sync.WaitGroup
+	for _, file := range deltas {
+		wg.Add(1)
+		go func(file *File, statedir string) {
+			defer wg.Done()
+			err := createDelta(file, statedir, file.DeltaPeer.Version, file.DeltaPeer.Hash)
+			if err != nil {
+				errchan <- file
+			}
+		}(file, statedir)
+	}
+
+	go func() {
+		wg.Wait()
+		close(errchan)
+	}()
+
+	for f := range errchan {
+		failed = append(failed, f)
+	}
+
+	return failed, nil
+}
+
+func consolidateDeltaFiles(manifest *Manifest, c config) ([]*File, error) {
+	var files []*File
+
+	if manifest == nil {
+		return nil, nil
+	}
+
+	for _, file := range manifest.Files {
+		if file.Version <= manifest.Header.Version ||
+			file.DeltaPeer == nil ||
+			(file.Type != typeFile && file.Type != typeDirectory && file.Type != typeLink) {
+			continue
+		}
+
+		deltaFile := fmt.Sprintf("%s/%d/delta/%d-%d-%s-%s",
+			c.outputDir, file.Version, file.DeltaPeer.Version,
+			file.Version, file.DeltaPeer.Hash, file.Hash)
+
+		// Only add to list if the delta file does not already exist on the
+		// system or in the list
+		if _, err := os.Stat(deltaFile); os.IsNotExist(err) {
+			if found := file.findFileNameInSlice(files); found == nil {
+				// Prepend to list...
```

Comment:
Can this be changed to just regular append()?

Reply:
Since you are doing nothing else with these files except returning them and don't rely on them to be sorted this can be a regular append, which will be much faster.